Add explicit M1 sources

ffoucart opened the pull request:
## Proposed changes

Adds computation of the explicit source terms for M1 radiation transport.

### Types of changes:

- [ ] Bugfix
- [ x] New feature

### Component:

- [ x] Code
- [ ] Documentation
- [ ] Build system
- [ ] Continuous integration

### Code review checklist

- [ ] The PR passes all checks, including unit tests, `clang-tidy` and `IWYU`.
  For instructions on how to perform the CI checks locally refer to the [Dev
  guide on the Travis CI](https://spectre-code.org/travis_guide.html).
- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).

### Further comments

Adds code computing the part of the source terms for the M1 code that are typically treated explicitly. The matter coupling terms have to be treated implicitly, so we'll have to discuss those later.

Open question: how do we want to treat multiple species of neutrinos? We should be able to call the current 'apply' function for each species, looping over all neutrino species (using a list provided in the Metavariables, as proposed by @nilsdeppe). I'm not sure what the best way to do this in a manner consistent with a future M1 executable would be though.  

Note also that the current function assumes that we know the pressure tensor. We need a separate function coding the M1 closure (provides P^ij as a function of E, F_i, and the metric).

moxcodes committed:
Add ComplexDataView

- This class provides a reference into a contiguous block of complex data, such
  that the real and imaginary parts of the complex data can easily be mutated or
  retrieved individually while minimizing copies.
- The option between two possible representations of complex numbers is
  provided, either alternating real and imaginary parts or a pair of contiguous
  real and contiguous imaginary blocks.

wthrowe committed:
Use new ordering functionals in FinalTime

wthrowe committed:
Use new ordering functionals in PastTime

wthrowe committed:
Use new ordering functionals in AdamsBashforthN

wthrowe committed:
Use new ordering functionals in TimeId

wthrowe committed:
Add ordering functionals that respect time flow

ffoucart committed:
Add explicit M1 sources

hrueter commented on src/Evolution/Systems/RadiationTransport/M1Grey/Sources.hpp:
```diff
@@ -0,0 +1,56 @@
+ * S({\tilde E}) = & \alpha \tilde P^{ij} K_{ij} - \tilde F^i \partial_i \alpha,
+ * S({\tilde F_i}) = & -\tilde E \partial_i \alpha + \tilde F_k \partial_i
+ * \beta^k
+ * + \frac{1}{2} \lapse \tilde P^{ij} \partial_i \gamma_{jk},
```
I guess instead of `\lapse` you actually meant to write `\alpha`.

ffoucart replied:
Indeed, thanks.

kidder commented on src/Evolution/Systems/RadiationTransport/M1Grey/Sources.hpp:
```diff
@@ -0,0 +1,56 @@
+
+/// \cond
+class DataVector;
+
```
`/// \endcond`

kidder commented on src/Evolution/Systems/RadiationTransport/M1Grey/Sources.hpp:
```diff
@@ -0,0 +1,56 @@
+ * grey M1 radiation transport.
+ *
+ *
+ * A flux-conservative system has the generic form:
```
This my fault, but it is a flux-balanced system.  (Mathematicians reserve flux-conservative systems to have no sources)

kidder commented on src/Evolution/Systems/RadiationTransport/M1Grey/Sources.hpp:
```diff
@@ -0,0 +1,56 @@
+ * + \frac{1}{2} \lapse \tilde P^{ij} \partial_i \gamma_{jk},
+ * \f}
+ *
+ * where \f${\tilde E}\f$, \f${\tilde F_i}\f$, \f${\tilde P}^{ij}\f$ are the
```
Can we use a different symbol than F for the densitized momentum (we are already using it for the flux) perhaps S for consistency with the fluid variables

ffoucart replied:
I'm ok either way. Using tildeF is consistent with the M1 literature, but on the other hand tildeF and tildeS are the exact same projection of the stress-energy tensor (one for the MHD part, the other for the radiation part), so there is certainly some value to using the same notation. 
If we do that, we may want to also use uniform notation for the pressure tensor  (S^{ij} in the fluid notation, P^{ij} in the neutrino notation), for consistency.

kidder commented on src/Evolution/Systems/RadiationTransport/M1Grey/Tags.hpp:
```diff
@@ -40,6 +40,17 @@ struct TildeF : db::SimpleTag {
 
+/// The densitized pressure tensor of neutrinos of a given species
+/// \f${\tilde P^{ij}}\f$
+/// computed from tildeE, tildeF using the M1 closure
```
latex tildeE, tildeF

kidder commented on src/Evolution/Systems/RadiationTransport/M1Grey/Sources.hpp:
```diff
@@ -0,0 +1,56 @@
+ * \f$K^{ij}\f$ is the extrinsic curvature, and \f$\alpha\f$, \f$\beta^i\f$,
+ * \f$\gamma^{ij}\f$ are the lapse, shift and 3-metric.
+ */
+static void apply(
```
Please make this a struct over all neutrino species, with the following apply function that calls your function which should be in a `detail` namespace and renamed `compute_sources_impl`
```cpp
template <typename... NeutrinoSpecies>
struct ComputeSources {
  static void apply(const gsl::not_null<db::item_type<NeutrinoSpecies>*>... sources_tilde_e,
    const gsl::not_null<db::item_type<TildeF<NeutrinoSpecies>>*>... sources_tilde_f,
    const db::item_type<NeutrinoSpecies>&... tilde_e,
   /* the remaining args */) noexcept {
       EXPAND_PACK_LEFT_TO_RIGHT(
          detail::compute_sources_impl(sources_tilde_e, sources_tilde_f, tilde_e, 
             /*the remaining args*/));
   }
};
```

ffoucart replied:
After the "EXPAND_PACK_LEFT_TO_RIGHT" command, what's the structure of, e.g., sources_tilde_e? i.e. how do I access each species after that command?

Edit: Never mind this comment, I hadn't seen the call to the 'detail' function inside EXPAND.

kidder reviewed (changes requested):
Reviewed with @nilsdeppe 
Please add a test.  If there is an analytic solution, you can test with it.  Also you can use the python testing infrastructure, see e.g. `tests/Unit/Evolution/Systems/GrMhd/ValenciaDivClean/Test_Sources.cpp`

ffoucart commented:
Thanks, I'll add a test once I have the loop working.

ffoucart commented:
Changed tildeF->tildeS and included the source terms for multiple species.
Could you check that "ComputeSources" is what you wanted me to do?

As a test, I can probably compute the source terms in the optically thin limit (S^2 = E^2, P^{ij} = S^i S^j / E), to avoid using the closure. I don't have an analytic solution, but can certainly compare with a python function, e.g. after choosing a random direction for the flux and a point in a Kerr background.

kidder commented on src/Evolution/Systems/RadiationTransport/M1Grey/Tags.hpp:
```diff
@@ -31,12 +31,12 @@ struct TildeE : db::SimpleTag {
+    return Frame::prefix<Fr>() + "TildeS_" + neutrinos::get_name(Species{});
   }
 };
 
```
There is one more TildeF -> TildeS in the description of TildeP

kidder reviewed (changes requested):
the suggested tests sound good, and the ComputeSources looks good

ffoucart commented:
Ok, putting the test together should be the only thing left for this PR then (besides whatever more in-depth turn review may turn up, if needed).

ffoucart commented:
I tried to setup a test for the source, but run into a linking error that I have not so far been able to solve (see below - undefined reference to a function that, as far as I can tell, should exist).
[I removed the static qualifier for that function as it was giving a warning, but the error itself is the same with or without the qualifier]

Linking error (clang, release, wheeler):

../../lib/libTest_M1Grey.a(Test_Sources.cpp.o): In function RadiationTransport::M1Grey::ComputeSources<neutrinos::ElectronNeutrinos<0ul> >::apply(gsl::not_null<Tensor<DataVector, brigand::list<>, brigand::list<> >*>, gsl::not_null<Tensor<DataVector, brigand::list<brigand::integral_constant<int, 1> >, brigand::list<Tensor_detail::TensorIndexType<3ul, (UpLo)1, Frame::Inertial, (IndexType)0> > >*>, Tensor<DataVector, brigand::list<>, brigand::list<> > const&, Tensor<DataVector, brigand::list<brigand::integral_constant<int, 1> >, brigand::list<Tensor_detail::TensorIndexType<3ul, (UpLo)1, Frame::Inertial, (IndexType)0> > > const&, Tensor<DataVector, brigand::list<brigand::integral_constant<int, 1>, brigand::integral_constant<int, 1> >, brigand::list<Tensor_detail::TensorIndexType<3ul, (UpLo)0, Frame::Inertial, (IndexType)0>, Tensor_detail::TensorIndexType<3ul, (UpLo)0, Frame::Inertial, (IndexType)0> > > const&, Tensor<DataVector, brigand::list<>, brigand::list<> > const&, Tensor<DataVector, brigand::list<brigand::integral_constant<int, 1> >, brigand::list<Tensor_detail::TensorIndexType<3ul, (UpLo)1, Frame::Inertial, (IndexType)0> > > const&, Tensor<DataVector, brigand::list<brigand::integral_constant<int, 2>, brigand::integral_constant<int, 1> >, brigand::list<Tensor_detail::TensorIndexType<3ul, (UpLo)1, Frame::Inertial, (IndexType)0>, Tensor_detail::TensorIndexType<3ul, (UpLo)0, Frame::Inertial, (IndexType)0> > > const&, Tensor<DataVector, brigand::list<brigand::integral_constant<int, 2>, brigand::integral_constant<int, 1>, brigand::integral_constant<int, 1> >, brigand::list<Tensor_detail::TensorIndexType<3ul, (UpLo)1, Frame::Inertial, (IndexType)0>, Tensor_detail::TensorIndexType<3ul, (UpLo)1, Frame::Inertial, (IndexType)0>, Tensor_detail::TensorIndexType<3ul, (UpLo)1, Frame::Inertial, (IndexType)0> > > const&, Tensor<DataVector, brigand::list<brigand::integral_constant<int, 1>, brigand::integral_constant<int, 1> >, brigand::list<Tensor_detail::TensorIndexType<3ul, (UpLo)0, Frame::Inertial, (IndexType)0>, Tensor_detail::TensorIndexType<3ul, (UpLo)0, Frame::Inertial, (IndexType)0> > > const&, Tensor<DataVector, brigand::list<brigand::integral_constant<int, 1>, brigand::integral_constant<int, 1> >, brigand::list<Tensor_detail::TensorIndexType<3ul, (UpLo)1, Frame::Inertial, (IndexType)0>, Tensor_detail::TensorIndexType<3ul, (UpLo)1, Frame::Inertial, (IndexType)0> > > const&)':
/home/ffoucart/spectre/src/Evolution/Systems/RadiationTransport/M1Grey/Sources.hpp:84: undefined reference to 'RadiationTransport::M1Grey::detail::compute_sources_impl(gsl::not_null<Tensor<DataVector, brigand::list<>, brigand::list<> >*>, gsl::not_null<Tensor<DataVector, brigand::list<brigand::integral_constant<int, 1> >, brigand::list<Tensor_detail::TensorIndexType<3ul, (UpLo)1, Frame::Inertial, (IndexType)0> > >*>, Tensor<DataVector, brigand::list<>, brigand::list<> > const&, Tensor<DataVector, brigand::list<brigand::integral_constant<int, 1> >, brigand::list<Tensor_detail::TensorIndexType<3ul, (UpLo)1, Frame::Inertial, (IndexType)0> > > const&, Tensor<DataVector, brigand::list<brigand::integral_constant<int, 1>, brigand::integral_constant<int, 1> >, brigand::list<Tensor_detail::TensorIndexType<3ul, (UpLo)0, Frame::Inertial, (IndexType)0>, Tensor_detail::TensorIndexType<3ul, (UpLo)0, Frame::Inertial, (IndexType)0> > > const&, Tensor<DataVector, brigand::list<>, brigand::list<> > const&, Tensor<DataVector, brigand::list<brigand::integral_constant<int, 1> >, brigand::list<Tensor_detail::TensorIndexType<3ul, (UpLo)1, Frame::Inertial, (IndexType)0> > > const&, Tensor<DataVector, brigand::list<brigand::integral_constant<int, 2>, brigand::integral_constant<int, 1> >, brigand::list<Tensor_detail::TensorIndexType<3ul, (UpLo)1, Frame::Inertial, (IndexType)0>, Tensor_detail::TensorIndexType<3ul, (UpLo)0, Frame::Inertial, (IndexType)0> > > const&, Tensor<DataVector, brigand::list<brigand::integral_constant<int, 2>, brigand::integral_constant<int, 1>, brigand::integral_constant<int, 1> >, brigand::list<Tensor_detail::TensorIndexType<3ul, (UpLo)1, Frame::Inertial, (IndexType)0>, Tensor_detail::TensorIndexType<3ul, (UpLo)1, Frame::Inertial, (IndexType)0>, Tensor_detail::TensorIndexType<3ul, (UpLo)1, Frame::Inertial, (IndexType)0> > > const&, Tensor<DataVector, brigand::list<brigand::integral_constant<int, 1>, brigand::integral_constant<int, 1> >, brigand::list<Tensor_detail::TensorIndexType<3ul, (UpLo)0, Frame::Inertial, (IndexType)0>, Tensor_detail::TensorIndexType<3ul, (UpLo)0, Frame::Inertial, (IndexType)0> > > const&, Tensor<DataVector, brigand::list<brigand::integral_constant<int, 1>, brigand::integral_constant<int, 1> >, brigand::list<Tensor_detail::TensorIndexType<3ul, (UpLo)1, Frame::Inertial, (IndexType)0>, Tensor_detail::TensorIndexType<3ul, (UpLo)1, Frame::Inertial, (IndexType)0> > > const&)'
collect2: error: ld returned 1 exit status
Fatal Error by charmc in directory /home/ffoucart/spectre/build_clang/tests/Unit
   Command g++ -D_REENTRANT -rdynamic -pthread -rdynamic -L/usr/local/charm/6.8.0/mpi-linux-x86_64-smp-gcc/lib -L/usr/lib64/ -L/usr/local/gcc/7.3.0/lib/gcc/x86_64-pc-linux-gnu/7.3.0 -L/usr/local/gcc/7.3.0/lib -Wl,-rpath,/usr/local/charm/6.8.0/mpi-linux-x86_64-smp-gcc/lib_so/ -o ../../bin/RunTests -L/usr/local/charm/6.8.0/mpi-linux-x86_64-smp-gcc/lib_so CMakeFiles/RunTests.dir/RunTests.cpp.o CMakeFiles/RunTests.dir/Test_ActionTesting.cpp.o CMakeFiles/RunTests.dir/Test_TestHelpers.cpp.o CMakeFiles/RunTests.dir/Test_TestingFramework.cpp.o CMakeFiles/RunTests.dir/Parallel/Test_ConstGlobalCacheDataBox.cpp.o CMakeFiles/RunTests.dir/Parallel/Test_GotoAction.cpp.o CMakeFiles/RunTests.dir/Parallel/Test_Parallel.cpp.o CMakeFiles/RunTests.dir/Parallel/Test_ParallelComponentHelpers.cpp.o CMakeFiles/RunTests.dir/Parallel/Test_PupStlCpp11.cpp.o CMakeFiles/RunTests.dir/Parallel/Test_TypeTraits.cpp.o /usr/local/python/anaconda2-4.1.1/lib/libpython2.7.so /usr/local/openblas/0.2.18/lib/libopenblas.so /usr/local/boost/1.65.0-gcc-6.4.0/lib/libboost_program_options.so /usr/lib64/libgsl.so /usr/lib64/libgslcblas.so /usr/local/python/anaconda2-4.1.1/lib/libhdf5.so /usr/lib64/librt.so /usr/local/python/anaconda2-4.1.1/lib/libz.so /usr/lib64/libdl.so /usr/lib64/libm.so /usr/local/jemalloc/5.0.1/lib/libjemalloc.so /usr/local/openblas/0.2.18/lib/libopenblas.so /usr/local/openblas/0.2.18/lib/libopenblas.so /usr/local/yaml-cpp/master/lib/libyaml-cpp.so /usr/local/python/anaconda2-4.1.1/lib/libpython2.7.so /usr/local/boost/1.65.0-gcc-6.4.0/lib/libboost_program_options.so /usr/lib64/libgsl.so /usr/lib64/libgslcblas.so /usr/local/python/anaconda2-4.1.1/lib/libhdf5.so /usr/lib64/librt.so /usr/local/python/anaconda2-4.1.1/lib/libz.so /usr/lib64/libdl.so /usr/lib64/libm.so /usr/local/jemalloc/5.0.1/lib/libjemalloc.so /usr/local/yaml-cpp/master/lib/libyaml-cpp.so RunTests_InfoFromBuild.o moduleinit12625.o -lmoduleDistributedLB ../../lib/libTest_ApparentHorizons.a ../../lib/libTest_ControlSystem.a ../../lib/libTest_DataBox.a ../../lib/libTest_EagerMath.a ../../lib/libTest_Expressions.a ../../lib/libTest_Tensor.a ../../lib/libTest_DataStructures.a ../../lib/libTest_Amr.a ../../lib/libTest_CoordinateMaps.a ../../lib/libTest_DomainCreators.a ../../lib/libTest_Domain.a ../../lib/libTest_EllipticActions.a ../../lib/libTest_EllipticDG.a ../../lib/libTest_Poisson.a ../../lib/libTest_ThermalNoise.a ../../lib/libTest_Xcts.a ../../lib/libTest_ErrorHandling.a ../../lib/libTest_EvolutionActions.a ../../lib/libTest_EvolutionConservative.a ../../lib/libTest_SlopeLimiters.a ../../lib/libTest_EvolutionDiscontinuousGalerkin.a ../../lib/libTest_EvolutionEventsAndTriggers.a ../../lib/libTest_Burgers.a ../../lib/libTest_CurvedScalarWave.a ../../lib/libTest_GeneralizedHarmonic.a ../../lib/libTest_ValenciaDivClean.a ../../lib/libTest_NewtonianEulerSources.a ../../lib/libTest_NewtonianEuler.a ../../lib/libTest_M1Grey.a ../../lib/libTest_Valencia.a ../../lib/libTest_ScalarWave.a ../../lib/libTest_VariableFixing.a ../../lib/libTest_IO.a ../../lib/libTest_Informer.a ../../lib/libTest_NumericalFluxes.a ../../lib/libTest_NumericalDiscontinuousGalerkin.a ../../lib/libTest_NumericalInterpolation.a ../../lib/libTest_LinearAlgebra.a ../../lib/libTest_LinearOperators.a ../../lib/libTest_LinearSolverActions.a ../../lib/libTest_ConjugateGradient.a ../../lib/libTest_Gmres.a ../../lib/libTest_LinearSolver.a ../../lib/libTest_RootFinding.a ../../lib/libTest_Spectral.a ../../lib/libTest_Options.a ../../lib/libTest_GrMhdAnalyticData.a ../../lib/libTest_BurgersSolutions.a ../../lib/libTest_GeneralRelativitySolutions.a ../../lib/libTest_GrMhdSolutions.a ../../lib/libTest_NewtonianEulerSolutions.a ../../lib/libTest_PoissonSolutions.a ../../lib/libTest_RelativisticEulerSolutions.a ../../lib/libTest_WaveEquation.a ../../lib/libTest_GeneralRelativity.a ../../lib/libTest_Hydro.a ../../lib/libTest_MathFunctions.a ../../lib/libTest_Pypp.a ../../lib/libTest_TestUtilities.a ../../lib/libTest_Time.a ../../lib/libTest_Utilities.a -lgsl -lgslcblas -lpthread /usr/local/libxsmm/1.8.1/lib/libxsmm.a /usr/local/libsharp/1.0.0/lib/libsharp.a /usr/local/libsharp/1.0.0/lib/libfftpack.a /usr/local/libsharp/1.0.0/lib/libc_utils.a ../../lib/libApparentHorizons.a ../../lib/libSPHEREPACK.a ../../lib/libAmr.a ../../lib/libControlSystem.a ../../lib/libPoisson.a ../../lib/libSlopeLimiters.a ../../lib/libCurvedScalarWave.a ../../lib/libValenciaDivClean.a ../../lib/libNewtonianEulerSources.a ../../lib/libM1Grey.a ../../lib/libTest_Domain.a ../../lib/libDomainCreators.a ../../lib/libValencia.a ../../lib/libScalarWave.a ../../lib/libVariableFixing.a ../../lib/libIO.a ../../lib/libInformer.a ../../lib/libDiscontinuousGalerkin.a ../../lib/libLinearAlgebra.a ../../lib/libLinearSolver.a ../../lib/libRootFinding.a ../../lib/libGrMhdAnalyticData.a ../../lib/libBurgers.a ../../lib/libBurgersSolutions.a ../../lib/libGeneralizedHarmonic.a ../../lib/libGrMhdSolutions.a ../../lib/libNewtonianEulerSolutions.a ../../lib/libNewtonianEuler.a ../../lib/libPoissonSolutions.a ../../lib/libRelativisticEulerSolutions.a ../../lib/libGeneralRelativitySolutions.a ../../lib/libInterpolation.a ../../lib/libGeneralRelativity.a ../../lib/libWaveEquation.a ../../lib/libHydro.a ../../lib/libMathFunctions.a ../../lib/libTestUtilities.a ../../lib/libTime.a ../../lib/libUtilities.a ../../lib/libDataStructures.a ../../lib/libCoordinateMaps.a ../../lib/libDomain.a ../../lib/libSpectral.a ../../lib/libLinearOperators.a ../../lib/libDataStructures.a ../../lib/libCoordinateMaps.a ../../lib/libDomain.a ../../lib/libSpectral.a ../../lib/libLinearOperators.a ../../lib/libErrorHandling.a -lgsl -lgslcblas -lpthread /usr/local/libxsmm/1.8.1/lib/libxsmm.a /usr/local/libsharp/1.0.0/lib/libsharp.a /usr/local/libsharp/1.0.0/lib/libfftpack.a /usr/local/libsharp/1.0.0/lib/libc_utils.a -lgfortran -lm -lquadmath -lckmain -lck -lmemory-default -lthreads-default -lconv-cplus-y -lconv-core -ltmgr -lconv-util -lconv-partition -lmemory-default -lthreads-default -lldb-rand -lconv-ldb -lpthread -lckqt -ldl -Wl,-rpath,/usr/local/python/anaconda2-4.1.1/lib -lmoduleDistributedLB -lmoduleNDMeshStreamer -lmodulecompletion -lmpi -lm returned error code 1

ffoucart commented:
(For some reason, the pointer signs don't appear in my copy-past above, but the first two arguments of the functions apply and compute_sources_impl are pointers, as they should be).

wthrowe commented:
The cpp file calls the function "compute_source_impl", but the hpp file calls it "compute_source**s**_impl".

ffoucart commented:
Thanks! 
[And I only checked that the name was the same 10 times... guess I had reach the point when I wasn't really reading my own code anymore...]

ffoucart commented:
@kidder : this one should be ready for review as well [up to whatever IWYU / clang-tidy asks me to fix], but it may be better to leave it 'in progress' until we merge the M1Closure PR (the two PRs create some of the same files, so merge conflicts will have to be resolved).
Let me know if you think that I should open this PR for full review as well though.

wthrowe committed:
Clean up includes in FinalTime

mar-celine committed:
Improve Wedge3D documentation of halves_to_use

moxcodes committed:
Add type trait is_complex_or_fundamental_v

moxcodes committed:
Add is_derived_of_vector_impl

- type trait for working out if an object is a SpECTRE vector

moxcodes committed:
Eliminate get_used_for_size, add size overloads

... to the vector MakeWithValues

moxcodes committed:
Rework SpinWeighted to support weighted vectors

moxcodes committed:
Add SpinWeighted typetraits

- Use them to white-list Tensors of spin-weighted ComplexDataVectors
- Also allow Variables with Tags with various Tensors of spin-weighted
  ComplexDataVectors

mar-celine committed:
Fix memory issue 1018 in DomainHelpers

wthrowe committed:
Clean up includes in PastTime

wthrowe committed:
Remove unused objects from trigger test DataBoxes

prayush committed:
Add return-by-ref versions for derivs of lapse

prayush committed:
update doxygen reference

kidder committed:
Move building Test_DataStructures to Travis second stage

ffoucart commented:
Replaced by #1366 , to merge more cleanly with the M1 closure PR. To be deleted if no one wants to keep this discussion as reference for the new PR.

kidder commented:
I am closing this as #1366 is essentially done.
